pmd-java/src/test/kotlin/net/sourceforge/pmd/lang/java/ast/ASTMethodReferenceTest.kt

```diff
@@ -83,111 +80,117 @@ class ASTMethodReferenceTest : ParserTestSpec({
-            it::getTypeArguments shouldBe null
+        inContext(ExpressionParsingCtx) {
 
-            it::getLhsExpression shouldBe null
-            it::getLhsType shouldBe child<ASTClassOrInterfaceType> {
-                it::getImage shouldBe "b"
-                it::getTypeImage shouldBe "foobar.b"
+            "super::foo" should parseAs {
+                methodRef("foo") {
+                    it::getLhsType shouldBe null
+                    it::getLhsExpression shouldBe child<ASTSuperExpression> {
 
-                it::getAmbiguousLhs shouldBe child<ASTAmbiguousName> {
-                    it::getName shouldBe "foobar"
+                    }
                 }
             }
 
-        }
-
+            "T.B.super::foo" should parseAs {
+                methodRef("foo") {
+                    it::getLhsType shouldBe null
+                    it::getLhsExpression shouldBe child<ASTSuperExpression> {
+                        it::getQualifier shouldBe classType("B") {
+                            ambiguousName("T")
+                        }
+                    }
+                }
+            }
 
-        "foobar.b<B>::new" should matchExpr<ASTMethodReference> {
+        }
 
-            it::getImage shouldBe "new"
-            it::getMethodName shouldBe null
-            it::isConstructorReference shouldBe true
-            it::getTypeArguments shouldBe null
+    }
 
-            it::getLhsExpression shouldBe null
-            it::getLhsType shouldBe child<ASTClassOrInterfaceType> {
-                it::getTypeImage shouldBe "foobar.b"
-                it::getImage shouldBe "b"
+    parserTest("Neg tests") {
 
-                it::getAmbiguousLhs shouldBe child {
-                    it::getName shouldBe "foobar"
-                    it::getTypeImage shouldBe "foobar"
-                }
+        inContext(ExpressionParsingCtx) {
 
+            "foo::bar::bar" shouldNot parse()
+            "foo::bar.foo()" shouldNot parse()
+            "foo::bar.foo" shouldNot parse()
 
-                it::getTypeArguments shouldBe child {
-                    child<ASTClassOrInterfaceType> {
-                        it::getTypeImage shouldBe "B"
-                    }
-                }
-            }
         }
 
-        "int[]::new" should matchExpr<ASTMethodReference> {
+    }
 
-            it::getImage shouldBe "new"
-            it::getMethodName shouldBe null
-            it::isConstructorReference shouldBe true
-            it::getTypeArguments shouldBe null
+    parserTest("Constructor reference") {
 
-            it::getLhsExpression shouldBe null
-            it::getLhsType shouldBe child<ASTArrayType> {
-                it::getTypeImage shouldBe "int"
+        inContext(ExpressionParsingCtx) {
 
-                it::getElementType shouldBe child<ASTPrimitiveType> {
-                    it::getTypeImage shouldBe "int"
-                }
+            "foobar.b::new" should parseAs {
+                constructorRef {
```

Comment:
hm... how does this `constructorRef` test work? I'm probably missing something (`classType("x")`, `methodRef`, ...)

Reply:
ah... found it: TestExtensions.kt